### PR TITLE
Add kaiju_kaiju module

### LINKED ISSUE #1398: new module: kaiju
I think it would be good to have a module for [kaiju](https://bioconda.github.io/recipes/kaiju/README.html)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by talnor**
This PR adds a kaiju module for taxonomic classification. This is needed for [taxprofiler](https://github.com/nf-core/taxprofiler/issues/3). 

## PR checklist

Closes #1398 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/kaiju/kaiju/main.nf [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def input = meta.single_end ? "-i ${reads}" : "-i ${reads[0]} -j ${reads[1]}"
+    """
+    kaiju \\
+        $args \\
+        -z $task.cpus \\
+        -t ${dbnodes} \\
+        -f ${db} \\
+        -o ${prefix}.results.tsv \\
```
Comment (jfy133) [inline modules/kaiju/kaiju/main.nf:31]:
```suggestion
        -o ${prefix}.tsv \\
```

File: modules/kaiju/kaiju/main.nf [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/kaiju:1.8.2--h5b5514e_1':
+        'quay.io/biocontainers/kaiju:1.8.2--h5b5514e_1' }"
+
+    input:
+    tuple val(meta), path(reads)
+    tuple path(db), path(dbnodes)
+
+    output:
+    tuple val(meta), path('*.results.tsv'), emit: results
```
Comment (jfy133) [inline modules/kaiju/kaiju/main.nf:15]:
```suggestion
    tuple val(meta), path('*.tsv'), emit: results
```

File: modules/kaiju/kaiju/meta.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
+name: kaiju_kaiju
+description: Taxonomic classification of metagenomic sequence data using a protein reference database
+keywords:
+  - classify
+  - metagenomics
+  - fastq
+  - db
```
Comment (jfy133) [inline modules/kaiju/kaiju/meta.yml:7]:
```suggestion
 - taxonomic profiling
```

maybe?

File: modules/kaiju/kaiju/meta.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - results:
+      type: file
+      description: Results with taxonomic classification of each read
+      pattern: "*.results.tsv"
```
Comment (jfy133) [inline modules/kaiju/kaiju/meta.yml:48]:
```suggestion
      pattern: "*.tsv"
```

File: modules/kaiju/kaiju/main.nf (hunk 2 above)
Comment (jfy133) [reply modules/kaiju/kaiju/main.nf:15]: https://nf-co.re/developers/modules#naming-conventions

> Output file (and/or directory) names SHOULD just consist of only ${prefix} and the file-format suffix (e.g. ${prefix}.fq.gz or ${prefix}.bam).

File: modules/kaiju/kaiju/meta.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - reads:
+      type: file
+      description: |
+        List of input fastq/fasta files of size 1 and 2 for single-end and paired-end data,
+        respectively.
+      pattern: "*.{fastq,fastq.gz,fasta,fasta.gz}"
```
Comment (jfy133) [inline modules/kaiju/kaiju/meta.yml:28]: Can this also accept .fq/.fa/.fna/.fas etc?

**Review by jfy133 — COMMENTED** [review]
A few minor comments about input/output descriptions but otherwise LGTM!

File: modules/kaiju/kaiju/main.nf (hunk 2 above)
Comment (talnor) [reply modules/kaiju/kaiju/main.nf:15]: True, nice catch! That extra part is quite unnecessary without the kaiju2tables command.

File: modules/kaiju/kaiju/meta.yml (hunk 5 above)
Comment (talnor) [reply modules/kaiju/kaiju/meta.yml:28]: Yes, it should work with any type of extension for the fastq and fasta formats.

Comment (talnor) [reply modules/kaiju/kaiju/meta.yml:28]:
```suggestion
      pattern: "*.{fastq,fq,fasta,fa,fsa,fas,fna,fastq.gz,fq.gz,fasta.gz,fa.gz,fsa.gz,fas.gz,fna.gz}"
```

File: tests/modules/kaiju/kaiju/test.yml [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    - path: output/kaiju/versions.yml
+      md5sum: 7e218c0ea00a71dd3a5ec5aaf28804f4
+
+- name: kaiju kaiju test_kaiju_kaiju_paired_end
+  command: nextflow run tests/modules/kaiju/kaiju -entry test_kaiju_kaiju_paired_end -c tests/config/nextflow.config
+  tags:
+    - kaiju/kaiju
+    - kaiju
+  files:
+    - path: output/kaiju/test.results.tsv
```
Comment (talnor) [inline tests/modules/kaiju/kaiju/test.yml:18]:
```suggestion
    - path: output/kaiju/test.tsv
```

File: modules/kaiju/kaiju/meta.yml (hunk 5 above)
Comment (jfy133) [reply modules/kaiju/kaiju/meta.yml:28]: I guess although this indeed looks slight overkill :laughing: But better to be explicit :+1:

Comment (talnor) [reply modules/kaiju/kaiju/meta.yml:28]: Alright! Then I'll commit the overkill version instead :laughing:

**Review by jfy133 — APPROVED** [review]

**Pull request merged**